File: 2019/Python/Day2.py
"""
Topic   : Advent of Code
Year    : 2019
Day     : 2
Title   : 1202 Program Alarm
"""

import numpy as np
from numpy import loadtxt

# =============================================================================
# Part One
# =============================================================================
def Intcode(x):
    pos = 0
    while x[pos] != 99:
        pos1 = x[pos + 1]
        pos2 = x[pos + 2]
        pos3 = x[pos + 3]
        if x[pos] == 1:
            x[pos3] = x[pos1] + x[pos2]
        elif x[pos] == 2:
            x[pos3] = x[pos1] * x[pos2]
        pos += 4
    return x

print(Intcode([1,0,0,0,99]))
print(Intcode([2,3,0,3,99]))
print(Intcode([2,4,4,5,99,0]))
print(Intcode([1,1,1,4,99,5,6,0,99]))

programInput = loadtxt("Day2Input.txt", delimiter = ",").astype(int)
programInput[1] = 12
programInput[2] = 2
print(Intcode(programInput)[0])

# =============================================================================
# Part Two
# =============================================================================
programInput = loadtxt("Day2Input.txt", delimiter = ",").astype(int)
output = 19690720
for i in np.arange(100):
    for j in np.arange(100):
        x = np.concatenate(([programInput[0]], [i, j], programInput[3:]))
        if Intcode(x)[0] == output:
            print(100 * i + j)

File: 2019/R/Day2.R
# ============================================================================ #
# Topic : Advent of Code
# --- Day 2: 1202 Program Alarm ---
# ============================================================================ #
setwd("~/Advent of Code/2019/2")
input <- as.vector(read.csv("input.txt", header = F), mode = "numeric")
Intcode <- function(input, initialize) {
  input[2] <- initialize[1]
  input[3] <- initialize[2]
  opcode <- input[1]
  i = 1
  while (opcode != 99) {
    pos1 <- input[i + 1] + 1
    pos2 <- input[i + 2] + 1
    pos3 <- input[i + 3] + 1
    if (opcode == 1) {
      input[pos3] <- input[pos1] + input[pos2]
    } else if (opcode == 2) {
      input[pos3] <- input[pos1] * input[pos2]
    } else if (opcode != 1 & opcode !=2) {stop("Something went wrong")}
    i <- i + 4
    opcode <- input[i]
  }
  return(input[1])
}
Intcode(input, c(12, 2))

# ============================================================================ #
# --- Part Two ---
# ============================================================================ #
for(i in 0:99) {
  for (j in 0:99)
    if (Intcode(input, c(i, j)) == 19690720) print(100 * i + j)
}